carma_wm_ctrl/src/WMBroadcaster.cpp

```diff
@@ -2075,6 +2078,60 @@ lanelet::Lanelet  WMBroadcaster::createLinearInterpolatingLanelet(const lanelet:
   return lanelet::Lanelet(lanelet::utils::getId(), createLinearInterpolatingLinestring(left_front_pt, left_back_pt, increment_distance), createLinearInterpolatingLinestring(right_front_pt, right_back_pt, increment_distance));
 }
 
+void WMBroadcaster::updateUpcomingSGIntersectionIds(const lanelet::Lanelet cur_lanelet)
+{
+  uint16_t map_msg_intersection_id = 0;
+  uint16_t cur_signal_group_id = 0;
+  for(auto itr = sim_.signal_group_to_entry_lanelet_ids_.begin(); itr != sim_.signal_group_to_entry_lanelet_ids_.end(); itr++)
+  {
+    if(itr->second.find(cur_lanelet.id()) != itr->second.end())
+    {
+      cur_signal_group_id = itr->first;
+    }
+  }
+
+  auto intersections = cur_lanelet.regulatoryElementsAs<lanelet::SignalizedIntersection>();
+  if (intersections.empty())
+  {
+    // no match if any of the entry lanelet is not part of any intersection.
+    ROS_DEBUG_STREAM("NO matching intersection for current lanelet. lanelet id = " << cur_lanelet.id());
+  }
+  else
+  {
+    lanelet::Ids cur_route_lanelet_ids = current_route.route_path_lanelet_ids;
+    lanelet::Id intersection_id = lanelet::InvalId ;
+
+    //Identify intersection based on the exit lanelets.
+    for(auto itr = intersections.begin(); itr != intersections.end(); itr++)
+    {
+      lanelet::ConstLanelets exit_lls = itr->get()->getExitLanelets();
+      
+      for(auto inner_itr = exit_lls.begin(); inner_itr != exit_lls.end(); inner_itr++)
+      {
+        if(std::find(cur_route_lanelet_ids.begin(), cur_route_lanelet_ids.end() , inner_itr->id() ) != cur_route_lanelet_ids.end())
+        {
+          intersection_id = itr->get()->id();
+        }
+      }      
+    }
+
+    if(intersection_id != lanelet::InvalId)
+    {
+      for(auto itr = sim_.intersection_id_to_regem_id_.begin(); itr != sim_.intersection_id_to_regem_id_.end(); itr++)
+      {
+        if(itr->second == intersection_id)
+        {
+          map_msg_intersection_id = itr->first;
+        }
+      }
+    }
```

Comment:
We have getSignalizedIntersectionsAlongRoute function to reuse here.
https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/carma_wm/include/carma_wm/CARMAWorldModel.h#L232
If we find an intersection, use sim again to find its intersection_id from lanelet type of id. (not Lanelet primitive's id)